tests/python/test_ggui.py

```diff
@@ -127,6 +127,124 @@ def render():
 @pytest.mark.skipif(not _ti_core.GGUI_AVAILABLE, reason="GGUI Not Available")
 @test_utils.test(arch=supported_archs)
 def test_geometry_3d():
+    window = ti.ui.Window("test", (640, 480), show_window=False)
+    canvas = window.get_canvas()
+    scene = window.get_scene()
+    camera = ti.ui.Camera()
+    camera.position(0.0, 0.0, 1.5)
+    camera.lookat(0.0, 0.0, 0)
+    scene.set_camera(camera)
+
+    # simple particles
+    num_per_dim = 32
+    num_particles_0 = int(num_per_dim**3)
+    particles_positions_0 = ti.Vector.field(3, ti.f32, shape=num_particles_0)
+
+    @ti.kernel
+    def init_particles_0():
+        for x, y, z in ti.ndrange(num_per_dim, num_per_dim, num_per_dim):
+            i = x * (num_per_dim**2) + y * num_per_dim + z
+            gap = 0.01
+            particles_positions_0[i] = ti.Vector([-0.4, 0, 0.0], dt=ti.f32) + ti.Vector([x, y, z], dt=ti.f32) * gap
+
+    init_particles_0()
+
+    # particles with individual colors
+    num_per_dim = 32
+    num_particles_1 = int(num_per_dim**3)
+    particles_positions_1 = ti.Vector.field(3, ti.f32, shape=num_particles_1)
+    particles_colors_1 = ti.Vector.field(3, ti.f32, shape=num_particles_1)
+
+    @ti.kernel
+    def init_particles_1():
+        for x, y, z in ti.ndrange(num_per_dim, num_per_dim, num_per_dim):
+            i = x * (num_per_dim**2) + y * num_per_dim + z
+            gap = 0.01
+            particles_positions_1[i] = ti.Vector([0.2, 0, 0.0], dt=ti.f32) + ti.Vector([x, y, z], dt=ti.f32) * gap
+            particles_colors_1[i] = ti.Vector([x, y, z], dt=ti.f32) / num_per_dim
+
+    init_particles_1()
+
+    # mesh
+    vertices = ti.Vector.field(3, ti.f32, shape=8)
+    colors = ti.Vector.field(3, ti.f32, shape=8)
+
+    @ti.kernel
+    def init_mesh():
+        for i, j, k in ti.ndrange(2, 2, 2):
+            index = i * 4 + j * 2 + k
+            vertices[index] = ti.Vector([-0.1, -0.3, 0.0], dt=ti.f32) + ti.Vector([i, j, k], dt=ti.f32) * 0.25
+            colors[index] = ti.Vector([i, j, k], dt=ti.f32)
+
+    init_mesh()
+    indices = ti.field(ti.i32, shape=36)
+    indices_np = np.array(
+        [
+            0,
```

Comment:
lol did the formatter do this

Can you replace this with something procedural?